File: Ponykart/Actors/Components/ModelComponent.cs
﻿using Mogre;
using Ponykart.Core;
using Ponykart.Levels;
using PonykartParsers;

namespace Ponykart.Actors {
	/// <summary>
	/// Represents an ogre node and mesh
	/// </summary>
	public class ModelComponent : LDisposable {
		public SceneNode Node { get; protected set; }
		public Entity Entity { get; protected set; }
		public long ID { get; protected set; }
		public string Name { get; protected set; }
		AnimationState Animation;

		/// <summary>
		/// Creates a model component for a Thing.
		/// </summary>
		/// <param name="lthing">The Thing this component is attached to</param>
		/// <param name="template">The template from the Thing</param>
		/// <param name="block">The block we're creating this component from</param>
		public ModelComponent(LThing lthing, ThingBlock template, ModelBlock block, ThingDefinition def) {
			ID = IDs.New;
			var sceneMgr = LKernel.GetG<SceneManager>();

			Name = block.GetStringProperty("name", template.ThingName);

			// if we're static, set up the static geometry
			if (block.GetBoolProperty("static", false) || def.GetBoolProperty("static", false)) {
				LKernel.GetG<StaticGeometryManager>().Add(this, template, block);
				Entity = null;
			}
			else if (block.GetBoolProperty("instanced", false) || def.GetBoolProperty("instanced", false)) {
				LKernel.GetG<InstancedGeometryManager>().Add(this, template, block);
				Entity = null;
			}
			// otherwise continue as normal
			else {
				Node = lthing.RootNode.CreateChildSceneNode(Name + "Node" + ID);

				// position
				Node.Position = block.GetVectorProperty("position", Vector3.ZERO);
				// orientation
				Node.Orientation = block.GetQuatProperty("orientation", Quaternion.IDENTITY);
				// if orientation was not found, we fall back to rotation
				if (Node.Orientation == Quaternion.IDENTITY) {
					Vector3 rot = block.GetVectorProperty("rotation", Vector3.ZERO);
					if (rot != Vector3.ZERO)
						Node.Orientation = rot.DegreeVectorToGlobalQuaternion();
				}
				// scale
				Node.Scale(block.GetVectorProperty("scale", Vector3.UNIT_SCALE));

				Node.InheritScale = true;
				Node.InheritOrientation = true;
				Node.SetInitialState();

				// make our entity
				Entity = sceneMgr.CreateEntity(Name + "Entity" + ID, block.GetStringProperty("mesh", null));

				// material name
				string materialName = block.GetStringProperty("material", string.Empty);
				if (!string.IsNullOrWhiteSpace(materialName))
					Entity.SetMaterialName(materialName);

				// some other properties
				Entity.CastShadows = block.GetBoolProperty("CastsShadows", false);

				SetupAnimation(block, def);

				// then attach it to the node!
				Node.AttachObject(Entity);
			}
		}

		/// <summary>
		/// Only does simple animations for now
		/// </summary>
		protected void SetupAnimation(ModelBlock block, ThingDefinition def) {
			if (block.GetBoolProperty("animated", false)) {
				Animation = Entity.GetAnimationState(block.GetStringProperty("AnimationName", null));
				Animation.Loop = block.GetBoolProperty("AnimationLooping", true);
				Animation.Enabled = block.GetBoolProperty("AnimationEnabled", true);

				LKernel.GetG<AnimationManager>().Add(Animation);
			}
		}

		protected override void Dispose(bool disposing) {
			if (IsDisposed)
				return;

			// stop updating the animation if we have one
			if (disposing && Animation != null) {
				LKernel.GetG<AnimationManager>().Remove(Animation);
			}

			var sceneMgr = LKernel.GetG<SceneManager>();
			bool valid = LKernel.GetG<LevelManager>().IsValidLevel;

			if (Entity != null) {
				if (valid && disposing)
					sceneMgr.DestroyEntity(Entity);
				Entity.Dispose();
				Entity = null;
			}
			if (Node != null) {
				if (valid && disposing)
					sceneMgr.DestroySceneNode(Node);
				Node.Dispose();
				Node = null;
			}

			base.Dispose(disposing);
		}

		public override string ToString() {
			return Node.Name;
		}
	}
}

File: Ponykart++/Actors/Components/ModelComponent.cpp
#include "pch.h"
#include "Actors/LThing.h"
#include "Actors/InstancedGeometryManager.h"
#include "Actors/StaticGeometryManager.h"
#include "Actors/Components/ModelComponent.h"
#include "Core/IDs.h"
#include "Core/Options.h"
#include "Core/Animation/AnimationBlender.h"
#include "Core/Animation/AnimationManager.h"
#include "Kernel/LKernel.h"
#include "Misc/ogreExtensions.h"
#include "Thing/ThingDefinition.h"
#include "Thing/Blocks/ModelBlock.h"
#include "Thing/Blocks/ThingBlock.h"

using namespace std;
using namespace Ogre;
using namespace Extensions;
using namespace PonykartParsers;
using namespace Ponykart::Actors;
using namespace Ponykart::Core;
using namespace Ponykart::LKernel;

ModelComponent::ModelComponent(LThing* lthing, ThingBlock* thingTemplate, ModelBlock* block, ThingDefinition* def)
: 	spawnPosition(block->getVectorProperty("position", Vector3::ZERO)),
	spawnOrientation(initSpawnOrientation(block)),
	spawnScale(block->getVectorProperty("scale", Vector3::UNIT_SCALE))
{
	id = IDs::incremental();
	owner = lthing;
	auto sceneMgr = LKernel::gSceneManager;

	name = block->getStringProperty("name", thingTemplate->getThingName());

	ThingEnum shad = block->getEnumProperty("CastsShadows", ThingEnum::Some);
	// if we're static, set up the static geometry
	// don't set up static geometry if we want to cast shadows though, since static geometry doesn't work with shadows
	if ((block->getBoolProperty("static", false) || def->getBoolProperty("static", false))
		// make static if we never want shadows
		&& (shad == ThingEnum::None
		// or if the mesh has "some" shadows but we don't want any
			|| (shad == ThingEnum::Some && Options::shadowDetail == ShadowDetailOption::None)
		// or if the mesh has "many" shadows but we only want those with "some"
			|| (shad == ThingEnum::Many && Options::shadowDetail != ShadowDetailOption::Many)))
	{
		LKernel::getG<StaticGeometryManager>()->add(this, thingTemplate, block, def);
		entity = nullptr;
	}
	else if (block->getBoolProperty("instanced", false) || def->getBoolProperty("instanced", false))
	{
		LKernel::getG<InstancedGeometryManager>()->add(this, thingTemplate, block, def);
		entity = nullptr;
	}
	// for attachments
	else if (block->getBoolProperty("Attached", false))
	{
		setupEntity(sceneMgr, block);
		setupAnimation(block);

		string boneName = block->getStringProperty("AttachBone");
		int modelComponentID = block->getFloatProperty("AttachComponentID"); // TODO: BUG:? Check float->int conversion
		Quaternion offsetQuat = block->getQuatProperty("AttachOffsetOrientation", Quaternion::IDENTITY);
		Vector3 offsetVec = block->getVectorProperty("AttachOffsetPosition", Vector3::ZERO);

		lthing->getModelComponents()[modelComponentID]->getEntity()->attachObjectToBone(boneName, entity, offsetQuat, offsetVec);
	}
	// otherwise continue as normal
	else
	{
		ostringstream ss; ss<<name<<"Node"<<id;
		node = lthing->getRootNode()->createChildSceneNode(ss.str());

		node->setPosition(spawnPosition);
		node->setOrientation(spawnOrientation);
		node->setScale(spawnScale);

		node->setInheritScale(block->getBoolProperty("InheritScale", true));
		node->setInheritOrientation(block->getBoolProperty("InheritOrientation", true));
		node->setInitialState();

		// make our entity
		setupEntity(sceneMgr, block);

		setupAnimation(block);

		// then attach it to the node!
		node->attachObject(entity);
	}
}

Entity* ModelComponent::getEntity()
{
	return entity;
}

unsigned ModelComponent::getId() const
{
	return id;
}

string ModelComponent::getName() const
{
	return name;
}

Ogre::SceneNode* ModelComponent::getNode()
{
	return node;
}

const LThing* const ModelComponent::getOwner() const
{
	return owner;
}

Quaternion ModelComponent::initSpawnOrientation(const ModelBlock* const block) const
{
	Quaternion q; Vector3 v;
	return (q=block->getQuatProperty("orientation", Quaternion::IDENTITY),
		q==Quaternion::IDENTITY?((v=block->getVectorProperty("rotation", Vector3::ZERO),
			v!=Vector3::ZERO?(degreeVectorToGlobalQuaternion(v)):q)):q);
}

void ModelComponent::setupAnimation(ModelBlock* block)
{
	if (block->getBoolProperty("animated", false))
	{
		int numAnims = 0;
		auto it = entity->getAllAnimationStates()->getAnimationStateIterator();
		for(auto i = it.begin(); i!=it.end(); ++i) ++numAnims;
		if (numAnims == 1)
		{
			animationState = entity->getAnimationState(block->getStringProperty("AnimationName"));
			animationState->setLoop(block->getBoolProperty("AnimationLooping", true));
			animationState->setEnabled(true);

			LKernel::getG<AnimationManager>()->add(animationState);
		}
		else if (numAnims > 1)
		{
			animationBlender = new AnimationBlender(entity);
			animationBlender->init(block->getStringProperty("AnimationName"), block->getBoolProperty("AnimationLooping", true));

			LKernel::getG<AnimationManager>()->add(animationBlender);
		}
	}
}

void ModelComponent::setupEntity(SceneManager* sceneMgr, ModelBlock* block)
{
	// make a new one if it isn't created yet, clone an existing one
	string meshName = block->getStringProperty("mesh");
	if (sceneMgr->hasEntity(meshName))
	{
		ostringstream ss; ss<<meshName<<id;
		entity = sceneMgr->getEntity(meshName)->clone(ss.str());
	}
	else
		entity = sceneMgr->createEntity(meshName, meshName);

	auto ft = block->getFloatTokens();
	if (ft.find("renderingdistance")!=ft.end())
		entity->setRenderingDistance(block->getFloatProperty("RenderingDistance"));

	// material name
	string materialName = block->getStringProperty("material", string());
	if (!materialName.empty() && materialName.find_first_not_of(' ') != string::npos)
		entity->setMaterialName(materialName);

	// some other properties
	ThingEnum shad = block->getEnumProperty("CastsShadows", ThingEnum::Some);
	if (Options::shadowDetail == ShadowDetailOption::Many)
		entity->setCastShadows(shad == ThingEnum::Many || shad == ThingEnum::Some);
	else if (Options::shadowDetail == ShadowDetailOption::Some)
		entity->setCastShadows(shad == ThingEnum::Some);
	else
		entity->setCastShadows(false);
}
